File: views/accounts_payable.py
# File: views/accounts_payable.py (continued)

def get_aging_summary():
    as_of_date = datetime.now()
    vendors = Vendor.query.all()
    
    aging_summary = []
    for vendor in vendors:
        bills = Bill.query.filter_by(vendor_id=vendor.id).all()
        current = overdue_30 = overdue_60 = overdue_90 = 0
        
        for bill in bills:
            payments = VendorPayment.query.filter_by(bill_id=bill.id).all()
            total_paid = sum(payment.amount for payment in payments)
            remaining = bill.total_amount - total_paid
            
            if remaining > 0:
                days_overdue = (as_of_date - bill.due_date).days
                if days_overdue <= 0:
                    current += remaining
                elif days_overdue <= 30:
                    overdue_30 += remaining
                elif days_overdue <= 60:
                    overdue_60 += remaining
                else:
                    overdue_90 += remaining
        
        if current or overdue_30 or overdue_60 or overdue_90:
            aging_summary.append({
                'vendor_name': vendor.name,
                'current': CurrencyUtil.format_currency(current),
                '1-30_days': CurrencyUtil.format_currency(overdue_30),
                '31-60_days': CurrencyUtil.format_currency(overdue_60),
                '60+_days': CurrencyUtil.format_currency(overdue_90),
                'total': CurrencyUtil.format_currency(current + overdue_30 + overdue_60 + overdue_90)
            })
    
    return jsonify(aging_summary)

def get_payment_efficiency():
    end_date = datetime.now()
    start_date = end_date - timedelta(days=90)
    
    bills = Bill.query.filter(Bill.bill_date.between(start_date, end_date)).all()
    total_billed = sum(bill.total_amount for bill in bills)
    total_paid = sum(
        payment.amount 
        for bill in bills 
        for payment in bill.payments 
        if payment.date <= end_date
    )
    
    efficiency = (total_paid / total_billed * 100) if total_billed > 0 else 0
    
    return jsonify({
        'total_billed': CurrencyUtil.format_currency(total_billed),
        'total_paid': CurrencyUtil.format_currency(total_paid),
        'efficiency': f"{efficiency:.2f}%"
    })

File: static/js/accounts_payable.js
// File: static/js/accounts_payable.js

document.addEventListener('DOMContentLoaded', function() {
    const vendorFilter = document.getElementById('vendor-filter');
    const asOfDateFilter = document.getElementById('as-of-date');
    const applyFiltersBtn = document.getElementById('apply-filters');
    const apTable = document.getElementById('ap-table');
    const agingTable = document.getElementById('aging-table');

    let agingChart, statusChart;

    applyFiltersBtn.addEventListener('click', loadAPData);

    function loadAPData() {
        const vendorId = vendorFilter.value;
        const asOfDate = asOfDateFilter.value;

        fetch(`/accounts-payable/data?vendor_id=${vendorId}&as_of_date=${asOfDate}`)
            .then(response => response.json())
            .then(data => {
                populateAPTable(data);
                updateCharts(data);
            });

        fetch('/accounts-payable/aging-summary')
            .then(response => response.json())
            .then(data => {
                populateAgingTable(data);
            });

        fetch('/accounts-payable/payment-efficiency')
            .then(response => response.json())
            .then(data => {
                updatePaymentEfficiency(data);
            });
    }

    function populateAPTable(data) {
        const tbody = apTable.querySelector('tbody');
        tbody.innerHTML = '';
        data.forEach(row => {
            const tr = document.createElement('tr');
            tr.innerHTML = `
                <td>${row.bill_id}</td>
                <td>${row.vendor_name}</td>
                <td>${row.bill_date}</td>
                <td>${row.due_date}</td>
                <td>${row.total_amount}</td>
                <td>${row.remaining}</td>
                <td>${row.days_overdue}</td>
                <td>${row.status}</td>
            `;
            tbody.appendChild(tr);
        });
    }

    function populateAgingTable(data) {
        const tbody = agingTable.querySelector('tbody');
        tbody.innerHTML = '';
        data.forEach(row => {
            const tr = document.createElement('tr');
            tr.innerHTML = `
                <td>${row.vendor_name}</td>
                <td>${row.current}</td>
                <td>${row['1-30_days']}</td>
                <td>${row['31-60_days']}</td>
                <td>${row['60+_days']}</td>
                <td>${row.total}</td>
            `;
            tbody.appendChild(tr);
        });
    }

    function updateCharts(data) {
        const agingData = {
            current: 0,
            '1-30_days': 0,
            '31-60_days': 0,
            '60+_days': 0
        };

        const statusData = {
            'Current': 0,
            'Slightly Overdue': 0,
            'Overdue': 0,
            'Very Overdue': 0,
            'Severely Overdue': 0
        };

        data.forEach(row => {
            const amount = parseFloat(row.remaining.replace(/[^0-9.-]+/g,""));
            if (row.days_overdue <= 0) agingData.current += amount;
            else if (row.days_overdue <= 30) agingData['1-30_days'] += amount;
            else if (row.days_overdue <= 60) agingData['31-60_days'] += amount;
            else agingData['60+_days'] += amount;

            statusData[row.status] += amount;
        });

        updateAgingChart(agingData);
        updateStatusChart(statusData);
    }

    function updateAgingChart(data) {
        const ctx = document.getElementById('aging-chart').getContext('2d');
        if (agingChart) agingChart.destroy();
        agingChart = new Chart(ctx, {
            type: 'pie',
            data: {
                labels: ['Current', '1-30 Days', '31-60 Days', '60+ Days'],
                datasets: [{
                    data: Object.values(data),
                    backgroundColor: ['#4CAF50', '#FFC107', '#FF9800', '#F44336']
                }]
            },
            options: {
                responsive: true,
                title: {
                    display: true,
                    text: 'AP Aging'
                }
            }
        });
    }

    function updateStatusChart(data) {
        const ctx = document.getElementById('status-chart').getContext('2d');
        if (statusChart) statusChart.destroy();
        statusChart = new Chart(ctx, {
            type: 'bar',
            data: {
                labels: Object.keys(data),
                datasets: [{
                    label: 'Amount',
                    data: Object.values(data),
                    backgroundColor: '#2196F3'
                }]
            },
            options: {
                responsive: true,
                title: {
                    display: true,
                    text: 'AP by Status'
                },
                scales: {
                    yAxes: [{
                        ticks: {
                            beginAtZero: true
                        }
                    }]
                }
            }
        });
    }

    function updatePaymentEfficiency(data) {
        const efficiencyDiv = document.getElementById('payment-efficiency');
        efficiencyDiv.innerHTML = `
            <h3>Payment Efficiency (Last 90 Days)</h3>
            <p>Total Billed: ${data.total_billed}</p>
            <p>Total Paid: ${data.total_paid}</p>
            <p>Efficiency: ${data.efficiency}</p>
        `;
    }

    // Initial load
    loadAPData();
});